lib/src/cpp/utils/XMLManager.cpp

```diff
@@ -186,209 +184,353 @@ bool XMLManager::save_xml()
+                xercesc::XMLString::transcode(name.c_str()),
+                xercesc::XMLString::transcode(value.c_str()));
+        }
+    }
+}
+
+std::string XMLManager::get_node_value()
+{
+    if (xercesc::XMLString::transcode(last_node->getNodeValue()) != nullptr)
+    {
+        return xercesc::XMLString::transcode(last_node->getNodeValue());
     }
-    // Complex element Node
-    else if (node_list->getLength() == 1)
+    // Throw ElementNotFound exception
+    std::string exception_message = "Could not find value of element ";
+    exception_message += xercesc::XMLString::transcode(last_node->getNodeName());
+    throw ElementNotFound(exception_message + "\n");
+}
+
+std::string XMLManager::get_node_attribute_value(
+        const std::string& name)
+{
+    xercesc::DOMNode* attribute_node = last_node->getAttributes()->getNamedItem(
+        xercesc::XMLString::transcode(name.c_str()));
+    if (attribute_node != nullptr)
+    {
+        if (xercesc::XMLString::transcode(attribute_node->getNodeValue()) != nullptr)
+        {
+            return xercesc::XMLString::transcode(attribute_node->getNodeValue());
+        }
+    }
+
+    // Throw ElementNotFound exception
+    std::string exception_message = "Could not find attribute ";
+    exception_message += name + " in element ";
+    exception_message += xercesc::XMLString::transcode(last_node->getNodeName());
+    throw ElementNotFound(exception_message + "\n");
+}
+
+void XMLManager::get_node(
+        const std::string& tag_name,
+        const bool& create_if_not_existent)
+{
+    // Obtain list of nodes based on the target tag
+    xercesc::DOMNodeList* node_list = static_cast<xercesc::DOMElement*>(last_node)->getElementsByTagName(
+        xercesc::XMLString::transcode(tag_name.c_str()));
+    // Check for expected node
+    if (node_list->getLength() == 1)
     {
         // Return Node
-        return node_list->item(0);
+        last_node = node_list->item(0);
+        return;
     }
-    // MAP or LIST element node
+
+    // Check if node should be created if not found
+    if (create_if_not_existent)
+    {
+        // Save parent node
+        xercesc::DOMNode* parent_node = last_node;
+
+        // Create new node and save it as last node
+        last_node = static_cast<xercesc::DOMNode*>(doc->createElement(xercesc::XMLString::transcode(tag_name.c_str())));
+
+        // Append new node to parent node
+        parent_node->appendChild(last_node);
+    }
+    // Throw ElementNotFound exception
     else
     {
-        // Obtain node
-        xercesc::DOMNode* tag_node = nullptr;
-        bool found = false;
+        throw ElementNotFound("non-existent " + tag_name + " element\n");
+    }
+}
+
+void XMLManager::get_node(
+        const std::string& index,
+        const std::string& default_tag_name,
+        const bool& create_if_not_existent)
+{
+    // Exception message declaration
+    std::string exception_message = "";
```

Comment:
```suggestion
    std::string exception_message;
```

Default constructor already initializes the string as empty.